components/eam/src/chemistry/modal_aero/sox_cldaero_mod.F90

```diff
@@ -88,436 +87,328 @@ subroutine sox_cldaero_init
+    ! compute factors for partitioning aerosol mass gains among modes
+    ! the factors are proportional to the activated particle MR for each
+    ! mode, which is the MR of cloud drops "associated with" the mode
+    ! thus we are assuming the cloud drop size is independent of the
+    ! associated aerosol mode properties (i.e., drops associated with
+    ! Aitken and coarse sea-salt particles are same size)
+    ! qnum_c(n) = activated particle number MR for mode n (these are just
+    ! used for partitioning among modes, so don't need to divide by cldfrc)
+    !-------------------------------------------------------------------------
+    real(r8), intent(in) :: tmr(:)   ! tracer mixing ratio [vmr]
+    integer,  intent(in) :: loffset  ! # of tracers in the host model that are not part of MAM
+    real(r8), intent(out) :: faqgain_so4(ntot_amode)   ! factor of TMR among modes [fraction]
+
+    ! local variables
+    integer  :: imode, ll               ! index
+    real(r8) :: sumf                    ! total TMR for all modes [vmr]
+    real(r8) :: qnum_c(ntot_amode)      ! tracer mixing ratio [vmr]
+    real(r8), parameter :: small_value_10 = 1.e-10_r8
+
+    !-------------------------------------------------------------------------
+    sumf  = 0.0_r8
+    qnum_c(:)  = 0.0_r8
+    faqgain_so4(:) = 0.0_r8
+
+    do imode = 1, ntot_amode
+
+       ll = numptrcw_amode(imode) - loffset
+       if (ll > 0) qnum_c(imode) = max( 0.0_r8, tmr(ll) )
+
+       ! force qnum_c(n) to be positive for n=modeptr_accum or n=1
+       if (imode == modeptr_accum) qnum_c(imode) = max(small_value_10, qnum_c(imode))
+
+       ! faqgain_so4(n) = fraction of total so4_c gain going to mode n
+       ! these are proportional to the activated particle MR for each mode
+       if (lptr_so4_cw_amode(imode) > 0) then
+          faqgain_so4(imode) = qnum_c(imode)
+          sumf = sumf + faqgain_so4(imode)
        endif
-    end do
 
-    sflx(:)=0._r8
-    do k=1,pver
-       do i=1,ncol
-          sflx(i)=sflx(i)+dqdt_aqhprxn(i,k)*specmw_so4_amode/mbar(i,k) &
-               *pdel(i,k)/gravit ! kg SO4 /m2/s
-       enddo
-    enddo
-    call outfld( 'AQSO4_H2O2', sflx(:ncol), ncol, lchnk)
-    sflx(:)=0._r8
-    do k=1,pver
-       do i=1,ncol
-          sflx(i)=sflx(i)+dqdt_aqo3rxn(i,k)*specmw_so4_amode/mbar(i,k) &
-               *pdel(i,k)/gravit ! kg SO4 /m2/s
-       enddo
     enddo
-    call outfld( 'AQSO4_O3', sflx(:ncol), ncol, lchnk)
 
-  end subroutine sox_cldaero_update
+    ! at this point (sumf <= 0.0) only when all the faqgain_so4 are zero
+    if (sumf > 0.0_r8) faqgain_so4(:) = faqgain_so4(:)/sumf
+
+  end subroutine compute_aer_factor
+
+  !=============================================================================
+  subroutine update_tmr ( tmr, dqdt, dtime )
+    !-----------------------------------------------------------------------
+    ! basically it just makes sure the value is greater than zero
```

Comment:
This isn't correct, is it? This updates `tmr` by applying the tendencies over the time interval `dt`.

Reply:
aha, sorry. This was just copy-pasted from the subroutine update_tmr_nonzero below... will change it.